File: mm-php/src/io/metamessage/ir/Ast.php
<?php

namespace io\metamessage\ir;

class NodeType
{
    public const UNKNOWN = 0;
    public const OBJECT = 1;
    public const ARRAY = 2;
    public const VALUE = 3;
    public const DOC = 4;
    public const NULL = 5;

    public const LABEL_UNKNOWN = 'unknown';
    public const LABEL_OBJECT = 'object';
    public const LABEL_ARRAY = 'array';
    public const LABEL_VALUE = 'value';
    public const LABEL_DOC = 'doc';
    public const LABEL_NULL = 'null';

    public static function toString(int $nt): string
    {
        return match ($nt) {
            self::OBJECT => self::LABEL_OBJECT,
            self::ARRAY => self::LABEL_ARRAY,
            self::VALUE => self::LABEL_VALUE,
            self::DOC => self::LABEL_DOC,
            self::NULL => self::LABEL_NULL,
            default => self::LABEL_UNKNOWN,
        };
    }

    public static function parse(string $s): int
    {
        return match ($s) {
            self::LABEL_OBJECT => self::OBJECT,
            self::LABEL_ARRAY => self::ARRAY,
            self::LABEL_VALUE => self::VALUE,
            self::LABEL_DOC => self::DOC,
            self::LABEL_NULL => self::NULL,
            default => self::UNKNOWN,
        };
    }
}

interface Node
{
    public function getTag(): ?Tag;

    public function getType(): int;

    public function getPath(): string;

    public function setPath(string $path): void;
}

class Field
{
    public string $Key;
    public Node $Value;

    public function __construct(string $key = '', ?Node $value = null)
    {
        $this->Key = $key;
        if ($value !== null) {
            $this->Value = $value;
        }
    }
}

class NodeObject implements Node
{
    public array $Fields = [];
    public ?Tag $Tag = null;
    public string $Path = '';

    public function getPath(): string
    {
        return $this->Path;
    }

    public function setPath(string $path): void
    {
        $this->Path = $path;
    }

    public function getType(): int
    {
        return NodeType::OBJECT;
    }

    public function getTag(): ?Tag
    {
        return $this->Tag;
    }
}

class NodeArray implements Node
{
    public array $Items = [];
    public ?Tag $Tag = null;
    public string $Path = '';

    public function getPath(): string
    {
        return $this->Path;
    }

    public function setPath(string $path): void
    {
        $this->Path = $path;
    }

    public function getType(): int
    {
        return NodeType::ARRAY;
    }

    public function getTag(): ?Tag
    {
        return $this->Tag;
    }
}

class NodeScalar implements Node
{
    public mixed $Data = null;
    public string $Text = '';
    public ?Tag $Tag = null;
    public string $Path = '';

    public function __construct(mixed $data = null, string $text = '', ?Tag $tag = null, string $path = '')
    {
        $this->Data = $data;
        $this->Text = $text;
        $this->Tag = $tag;
        $this->Path = $path;
    }

    public function getPath(): string
    {
        return $this->Path;
    }

    public function setPath(string $path): void
    {
        $this->Path = $path;
    }

    public function getType(): int
    {
        return NodeType::VALUE;
    }

    public function getTag(): ?Tag
    {
        return $this->Tag;
    }
}

class NodeNull implements Node
{
    public ?Tag $Tag = null;
    public string $Path = '';

    public function getPath(): string
    {
        return $this->Path;
    }

    public function setPath(string $path): void
    {
        $this->Path = $path;
    }

    public function getType(): int
    {
        return NodeType::NULL;
    }

    public function getTag(): ?Tag
    {
        return $this->Tag;
    }
}

class Doc implements Node
{
    public array $Fields = [];
    public ?Tag $Tag = null;
    public string $Path = '';

    public function getPath(): string
    {
        return $this->Path;
    }

    public function setPath(string $path): void
    {
        $this->Path = $path;
    }

    public function getType(): int
    {
        return NodeType::DOC;
    }

    public function getTag(): ?Tag
    {
        return $this->Tag;
    }
}

File: internal/ir/ast.go
package ir

type NodeType uint8

const (
	NodeTypeUnknown NodeType = iota
	NodeTypeDoc
	NodeTypeObject
	NodeTypeArray
	NodeTypeScalar
	NodeTypeNull

	LabelNodeTypeUnknown = "unknown"
	LabelNodeTypeDoc     = "doc"
	LabelNodeTypeObject  = "object"
	LabelNodeTypeArray   = "array"
	LabelNodeTypeScalar  = "scalar"
	LabelNodeTypeNull    = "null"
)

func (nt NodeType) String() string {
	switch nt {
	case NodeTypeDoc:
		return LabelNodeTypeDoc
	case NodeTypeObject:
		return LabelNodeTypeObject
	case NodeTypeArray:
		return LabelNodeTypeArray
	case NodeTypeScalar:
		return LabelNodeTypeScalar
	case NodeTypeNull:
		return LabelNodeTypeNull
	default:
		return LabelNodeTypeUnknown
	}
}

func ParseNodeType(s string) NodeType {
	switch s {
	case LabelNodeTypeDoc:
		return NodeTypeDoc
	case LabelNodeTypeObject:
		return NodeTypeObject
	case LabelNodeTypeArray:
		return NodeTypeArray
	case LabelNodeTypeScalar:
		return NodeTypeScalar
	case LabelNodeTypeNull:
		return NodeTypeNull
	default:
		return NodeTypeUnknown
	}
}

type Node interface {
	GetTag() *Tag
	GetType() NodeType

	GetPath() string
	SetPath(path string)
}

type Field struct {
	Key   string
	Value Node
}

type NodeObject struct {
	Fields []*Field
	Tag    *Tag
	Path   string
}

func (o *NodeObject) GetPath() string { return o.Path }

func (o *NodeObject) SetPath(path string) { o.Path = path }

func (o *NodeObject) GetType() NodeType { return NodeTypeObject }

func (o *NodeObject) GetTag() *Tag {
	if o == nil {
		return nil
	}
	if o.Tag != nil {
		return o.Tag
	}
	return nil
}

type NodeArray struct {
	Items []Node
	Tag   *Tag
	Path  string
}

func (a *NodeArray) GetPath() string { return a.Path }

func (a *NodeArray) SetPath(path string) { a.Path = path }

func (a *NodeArray) GetType() NodeType { return NodeTypeArray }

func (a *NodeArray) GetTag() *Tag {
	if a == nil {
		return nil
	}
	if a.Tag != nil {
		return a.Tag
	}
	return nil
}

type NodeScalar struct {
	Data any
	Text string
	Tag  *Tag
	Path string
}

func (v *NodeScalar) GetPath() string { return v.Path }

func (v *NodeScalar) SetPath(path string) { v.Path = path }

func (v *NodeScalar) GetType() NodeType { return NodeTypeScalar }

func (v *NodeScalar) GetTag() *Tag {
	if v == nil {
		return nil
	}
	if v.Tag != nil {
		return v.Tag
	}
	return nil
}

type Doc struct {
	Fields []*Field
	Tag    *Tag
	Path   string
}

func (d *Doc) GetPath() string { return d.Path }

func (d *Doc) SetPath(path string) { d.Path = path }

func (d *Doc) GetType() NodeType { return NodeTypeDoc }

func (d *Doc) GetTag() *Tag {
	if d == nil {
		return nil
	}
	if d.Tag != nil {
		return d.Tag
	}
	return nil
}

type NodeNull struct {
	Tag *Tag
}

func (d *NodeNull) GetPath() string { return "" }

func (d *NodeNull) SetPath(path string) {}

func (d *NodeNull) GetType() NodeType { return NodeTypeNull }

func (d *NodeNull) GetTag() *Tag {
	if d == nil {
		return nil
	}
	if d.Tag != nil {
		return d.Tag
	}
	return nil
}
